atef/widgets/happi.py

```diff
@@ -0,0 +1,470 @@
+
+            copy_action = self.menu.addAction(f"&Copy: {index.data()}")
+            copy_action.triggered.connect(copy)
+
+        self.menu.exec_(self.happi_list_view.mapToGlobal(pos))
+
+    @property
+    def selected_device_widget(self) -> Union[HappiDeviceListView, HappiDeviceTreeView]:
+        """The selected device widget - either the list or tree view."""
+        if self.radio_by_name.isChecked():
+            return self.happi_list_view
+
+        return self.happi_tree_view
+
+    @QtCore.Slot(str)
+    def _category_changed(self, category: str):
+        """By-category category has changed."""
+        if self._tree_has_data and self._tree_current_category == category:
+            return
+
+        self._tree_current_category = category
+        self.happi_tree_view.group_by(category)
+        # Bugfix (?) otherwise this ends up in descending order
+        self.happi_tree_view.model().sort(0, QtCore.Qt.AscendingOrder)
+        self.radio_by_category.setChecked(True)
+        self._select_device_widget()
+
+    @QtCore.Slot()
+    def _select_device_widget(self):
+        """Switch between the list/table view."""
+        selected = self.selected_device_widget
+        for widget in (self.happi_tree_view, self.happi_list_view):
+            widget.setVisible(selected is widget)
+
+        if self.happi_tree_view.isVisible() and not self._tree_has_data:
+            self._tree_has_data = True
+            self.refresh_happi()
+
+    @QtCore.Slot()
+    def refresh_happi(self):
+        """Search happi again and update the widgets."""
+        def search():
+            # TODO/upstream: this is coupled with 'search' in the view
+            HappiViewMixin.search(self.selected_device_widget)
+
+        def update_gui():
+            # TODO/upstream: this is coupled with 'search' in the view
+            self.selected_device_widget._update_data()
+            self.button_refresh.setEnabled(True)
+
+        def report_error(ex: Exception):
+            logger.warning("Failed to update happi information: %s", ex, exc_info=ex)
+            self.button_refresh.setEnabled(True)
+
+        if self._client is None:
+            return
+        if self._search_thread is not None and self._search_thread.isRunning():
+            return
+
+        self.button_refresh.setEnabled(False)
+        self._search_thread = ThreadWorker(search)
+        self._search_thread.finished.connect(update_gui)
+        self._search_thread.error_raised.connect(report_error)
+        self._search_thread.start()
+
+    @property
+    def client(self) -> Optional[happi.Client]:
+        """The client to use for search."""
+        return self._client
+
+    @client.setter
+    def client(self, client: Optional[happi.Client]):
+        self._client = client
+        self.happi_tree_view.client = client
+        self.happi_list_view.client = client
+        self.refresh_happi()
+
+
+class HappiItemMetadataView(DesignerDisplay, QtWidgets.QWidget):
```

Comment:
This widget looks good too. My understanding is that the runtime intended usage is to set `item_name` to whichever device name I'd like to display the metadata for. Pretty solid.
We'll probably use this to provide much needed context information when the user is selecting devices.

Reply:
You've got the right idea, intended usage-wise. It may have more uses than just what we're considering at the moment, even. Happi metadata editor utility instead of our poor excuse for an update CLI?